README.md
# Ronin Dorks

* http://ronin.rubyforge.org/dorks/
* http://github.com/postmodern/ronin-dorks
* http://github.com/postmodern/ronin-dorks/issues
* http://groups.google.com/group/ronin-ruby
* irc.freenode.net #ronin

## DESCRIPTION:

Ronin Dorks is a Ruby library for Ronin that provides support for various
Google (tm) Dorks functionality.

Ronin is a Ruby platform for exploit development and security research.
Ronin allows for the rapid development and distribution of code, exploits
or payloads over many common Source-Code-Management (SCM) systems.

### Ruby

Ronin's Ruby environment allows security researchers to leverage Ruby with
ease. The Ruby environment contains a multitude of convenience methods
for working with data in Ruby, a Ruby Object Database, a customized Ruby
Console and an extendable command-line interface.

### Extend

Ronin's more specialized features are provided by additional Ronin
libraries, which users can choose to install. These libraries can allow
one to write and run Exploits and Payloads, scan for PHP vulnerabilities,
perform Google Dorks  or run 3rd party scanners.

### Publish

Ronin allows users to publish and share code, exploits, payloads or other
data via Overlays. Overlays are directories of code and data that can be
hosted on any SVN, Hg, Git or Rsync server. Ronin makes it easy to create,
install or update Overlays.

## FEATURES/PROBLEMS:

* Provides convenience methods for many common Google (tm) Dorks:
  * Search for URLs containing a specified string.
  * Search for URLs containing a specified pattern.
  * Search for Pages containing a specified text string.
  * Search for Pages containing a specified text pattern.
  * Search for Indexes of unprotected `cgi-bin` directories.
  * Search for unprotected MySQL dump files.
  * Search for unprotected MySQL dump files containing the password hash
    'admin'.
  * Search for Certificate Practice Statement documents.
  * Search for Network Vulnerability Assessemnt reports.
  * Search for order receipts.
  * Search for `robots.txt` files.
  * Search for unprotected PHP MyAdmin panels.
  * Search for `emails.xls` files.
  * Search for `finances.xls` files.

## SYNOPSIS:

Start the Ronin console with Ronin Dorks preloaded:

    $ ronin dorks

## EXAMPLES:

Search for URLs containing a pattern:

    Web::Dorks.inurl('stmt=')

Search for URLs containing all patterns:

    Web::Dorks.allinurl(['show', 'php', 'page'])

Search for pages containing the specified text in the title:

    Web::Dorks.intitle('Profile')

Search for pages containing all specified text in the title:

    Web::Dorks.allintitle(['Profile', 'Edit'])

Search for pages containing the specified text:

    Web::Dorks.intext('Powered by')

Search for pages containing all specified text:

    Web::Dorks.allintext(['Powered by', '0.5'])

## REQUIREMENTS:

* [ronin](http://ronin.rubyforge.org/) >= 0.3.0
* [gscraper](http://gscraper.rubyforge.org/) >= 0.2.2

## INSTALL:

    $ sudo gem install ronin-dorks

## LICENSE:

Ronin Dorks - A Ruby library for Ronin that provides support for various
Google (tm) Dorks and Hacking functionality.

Copyright (c) 2008-2009 Hal Brodigan (postmodern.mod3 at gmail.com)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA


History.txt
=== 0.1.3 / 2009-09-24

* Require ronin >= 0.3.0.
* Require rspec >= 1.2.8.
* Require yard >= 0.2.3.5.
* Moved to YARD based documentation.
* Removed requirement on ronin-web.
* Updated the project summary and 3-point description for Ronin Dorks.

=== 0.1.2 / 2009-07-02

* Use Hoe >= 2.0.0.
* Added Web::Dorks.intext.
* Added Web::Dorks.allintext.
* Added Web::Dorks.string_intext.
* Added Web::Dorks.all_strings_intext.
* Added Web::Dorks.intitle.
* Added Web::Dorks.allintitle.
* Added Web::Dorks.string_intitle.
* Added Web::Dorks.all_strings_intitle.

=== 0.1.1 / 2009-02-23

* Added a git style sub-command (<tt>ronin-dorks</tt>) which starts the
  Ronin console with <tt>ronin/dorks</tt> pre-loaded.
* Added Dorks.all_strings_inurl.
* Added Dorks.index_of.
* Added Dorks.index_with_file.
* Added Dorks.index_with_sql.
* Added Dorks.sharepoint.
* Require GScraper >= 0.2.2.
* Require Ronin >= 0.2.1.
* Require Ronin Web >= 0.1.1.
* Renamed Dorks.mysql_dump to Dorks.sql_dump.
* Renamed Dorks.mysql_dump_admin to Dorks.sql_admin_dump.
* Fixed many small bugs.
* Updated README.txt.
* Added RSpec tests.

=== 0.1.0 / 2008-08-21

* Initial release.
* Provides convenience methods for many common Google (tm) Dorks:
  * Search for URLs containing a specified string:
  * Search for URLs containing a specified pattern:
  * Search for Indexes of unprotected <tt>cgi-bin</tt> directories.
  * Search for unprotected MySQL dump files.
  * Search for unprotected MySQL dump files containing the password hash
    'admin'.
  * Search for Certificate Practice Statement documents.
  * Search for Network Vulnerability Assessemnt reports.
  * Search for order receipts.
  * Search for <tt>robots.txt</tt> files.
  * Search for unprotected PHP MyAdmin panels.
  * Search for <tt>emails.xls</tt> files.
  * Search for <tt>finances.xls</tt> files.



lib/ronin/web/dorks/dorks.rb
#
# Ronin Dorks - A Ruby library for Ronin that provides support for various
# Google (tm) Dorks functionality.
#
# Copyright (c) 2008-2009 Hal Brodigan (postmodern.mod3 at gmail.com)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#

require 'gscraper/search'
require 'digest/md5'

module Ronin
  module Web
    module Dorks
      include GScraper

      #
      # Creates either a <tt>GScraper::Search::WebQuery</tt> or a
      # <tt>GScraper::Search::AJAXQuery</tt> with the given _options_.
      # If a _block_ is given, it will be passed the newly created 
      # query object.
      #
      # _options_ may contain the following keys:
      # <tt>:ajax</tt>:: Specifies wether or not to create a
      #                  <tt>GScraper::Search::AJAXQuery</tt>.
      # <tt>:query</tt>:: The search query.
      # <tt>:link</tt>:: Search for results which link to the specified
      #                  URI.
      # <tt>:related</tt>:: Search for results which relate to the
      #                     specified URI.
      # <tt>:info</tt>:: Return information about the specified URI.
      # <tt>:site</tt>:: Limit results to the specified site.
      # <tt>:filetype</tt>:: Limit results to those with the specified
      #                      file-type.
      # <tt>:allintitle</tt>:: Search for results with all of the keywords
      #                        appearing in the title.
      # <tt>:intitle</tt>:: Search for results with the keyword appearing
      #                     in the title.
      # <tt>:allintext</tt>:: Search for results with all of the keywords
      #                       appearing in the text.
      # <tt>:intext</tt>:: Search for results with the keyword appearing
      #                    in the text.
      # <tt>:allinanchor</tt>:: Search for results with all of the keywords
      #                         appearing in the text of links.
      # <tt>:inanchor</tt>:: Search for results with the keyword appearing
      #                      in the text of links.
      # <tt>:exact_phrase</tt>:: Search for results containing the specified
      #                          exact phrase.
      # <tt>:with_words</tt>:: Search for results containing all of the
      #                        specified words.
      # <tt>:without_words</tt>:: Search for results not containing any of
      #                           the specified words.
      # <tt>:numeric_range</tt>:: Search for results contain numbers that
      #                           fall within the specified Range.
      # <tt>:define</tt>:: Search for results containing the definition of
      #                    the specified keyword.
      #
      # If the <tt>:ajax</tt> option is not specified, then _options_
      # may contain the following additional keys:
      # <tt>:results_per_page</tt>:: Specifies the number of results
      #                              for each page.
      # <tt>:language</tt>:: Search for results in the specified language.
      # <tt>:region</tt>:: Search for results from the specified region.
      # <tt>:within_past_day</tt>:: Search for results that were created
      #                             within the past day.
      # <tt>:within_past_week</tt>:: Search for results that were created
      #                              within the past week.
      # <tt>:within_past_month</tt>:: Search for results that were created
      #                               within the past month.
      # <tt>:within_past_year</tt>:: Search for results that were created
      #                              within the past year.
      # <tt>:occurrs_within</tt>:: 
      # <tt>:rights</tt>:: Search for results licensed under the specified
      #                    license.
      # <tt>:filtered</tt>:: Specifies whether or not to use SafeSearch.
      #                      Defaults to +false+, if not specified.
      #
      # If the <tt>:ajax</tt> option is specified, then _options_
      # may contain the following additional keys:
      # <tt>:language</tt>:: The search language. Defaults to <tt>:en</tt>.
      # <tt>:sig</tt>:: The search signature. Defaults to
      #                 <tt>582c1116317355adf613a6a843f19ece</tt>.
      # <tt>:key</tt>:: The search key. Defaults to <tt>:notsupplied</tt>.
      # <tt>:version</tt>:: The desired API version. Defaults to
      #                     <tt>1.0</tt>.
      #
      def Dorks.search(options={},&block)
        if options[:ajax] == true
          return Search.ajax_query(options,&block)
        else
          return Search.query(options,&block)
        end
      end

      def Dorks.intext(pattern,options={},&block)
        Dorks.search(options.merge(:intext => pattern),&block)
      end

      def Dorks.allintext(pattern,options={},&block)
        Dorks.search(options.merge(:allintext => pattern),&block)
      end

      def Dorks.string_intext(string,options={},&block)
        Dorks.intext("'#{string}'",options,&block)
      end

      def Dorks.all_strings_intext(strings,options={},&block)
        Dorks.search(options.merge(:allintext => strings.map { |string|
          "'#{string}'"
        }),&block)
      end

      def Dorks.intitle(pattern,options={},&block)
        Dorks.search(options.merge(:intitle => pattern),&block)
      end

      def Dorks.string_intitle(string,options={},&block)
        Dorks.intitle("'#{string}'",options,&block)
      end

      def Dorks.allintitle(patterns,options={},&block)
        Dorks.search(options.merge(:allintitle => patterns),&block)
      end

      def Dorks.all_strings_intitle(strings,options={},&block)
        Dorks.search(options.merge(:allintitle => strings.map { |string|
          "'#{string}'"
        }),&block)
      end

      def Dorks.inurl(pattern,options={},&block)
        Dorks.search(options.merge(:inurl => pattern),&block)
      end

      def Dorks.string_inurl(string,options={},&block)
        Dorks.inurl("'#{string}'",options,&block)
      end

      def Dorks.allinurl(patterns,options={},&block)
        Dorks.search(options.merge(:allinurl => patterns),&block)
      end

      def Dorks.all_strings_inurl(strings,options={},&block)
        Dorks.search(options.merge(:allinurl => strings.map { |string|
          "'#{string}'"
        }),&block)
      end

      def Dorks.sql_dump(options={},&block)
        query = []
        
        query << "Host: #{options[:host]}" if options[:host]
        query << "Database: #{options[:database]}" if options[:database]
        query << options[:version] if options[:version]

        query << "\"#{options[:sql]}\"" if options[:sql]

        if options[:password]
          query << Digest::MD5.hexdigest(options[:password].to_s)
        end

        return Dorks.search(options.merge(
          :query => query,
          :exact_phrase => '"SQL Dump"',
          :filetype => :sql
        ),&block)
      end

      def Dorks.sql_admin_dump(options={},&block)
        Dorks.sql_dump(options.merge(:password => 'admin'),&block)
      end

      def Dorks.robots_txt(options={},&block)
        Dorks.search(options.merge(
          :exact_phrase => 'robots.txt',
          :with_words => ['Disallow'],
          :filetype => :txt
        ),&block)
      end

      def Dorks.php_my_admin(options={},&block)
        Dorks.search(options.merge(
          :with_words => ['phpMyAdmin'],
          :exact_phrase => 'running on',
          :inurl => 'main.php'
        ),&block)
      end

      def Dorks.download_file(options={},&block)
        Dorks.search(options.merge(
          :allinurl => ['download.php?', 'file']
        ),&block)
      end

      def Dorks.download_pdf(options={},&block)
        Dorks.search(options.merge(
          :allinurl => [
            'download.php?',
            'file',
            '.pdf'
          ]
        ),&block)
      end

      def Dorks.sharepoint(options={},&block)
        Dorks.search(options.merge(
          :exact_phrase => 'all site content'
        ),&block)
      end
    end
  end
end
